markdown-toc.el

```diff
@@ -272,15 +272,34 @@ the deleted TOC region, effectively replacing it."
   "Generate a TOC for markdown file at current point.
 Deletes any previous TOC."
   (interactive)
-  (save-excursion
-    (when (markdown-toc--toc-already-present-p)
-      ;; when toc already present, remove it
-      (markdown-toc--delete-toc t))
-    (->> (funcall imenu-create-index-function)
-         markdown-toc--compute-toc-structure
-         (funcall markdown-toc-user-toc-structure-manipulation-fn)
-         markdown-toc--generate-toc
-         insert)))
+  ;; Set truncate-lines to t because count-screen-lines becomes unreliable when
+  ;; truncate-lines is nil and there are many wrapped lines. In such cases, it
+  ;; may count more lines than are actually displayed on the screen.
+  (let* ((truncate-lines t)
+         (window-hscroll (window-hscroll))
+         (point (save-excursion (vertical-motion 0)
+                                (point)))
+         (window-start (save-excursion (goto-char (window-start))
+                                       (vertical-motion 0)
+                                       (point)))
+         (lines-before-marker (count-screen-lines window-start point)))
+    (unwind-protect
+        (save-excursion
+          (when (markdown-toc--toc-already-present-p)
+            ;; When toc already present, remove it
+            (markdown-toc--delete-toc t))
+          (->> (funcall imenu-create-index-function)
+               markdown-toc--compute-toc-structure
+               (funcall markdown-toc-user-toc-structure-manipulation-fn)
+               markdown-toc--generate-toc
+               insert))
+      (let ((window-start (save-excursion
+                            (vertical-motion 0)
+                            (line-move-visual (* -1 lines-before-marker))
+                            (vertical-motion 0)
+                            (point))))
+        (set-window-start (selected-window) window-start t))
+      (set-window-hscroll (selected-window) window-hscroll))))
```

Comment:
Why not use the macro `save-window-excursion`? 🤔

Reply:
Using `save-window-excursion` does not work because the buffer is modified by `markdown-toc`. This pull request restores `window-start` relative to the current line after the buffer has been modified by `markdown-toc`.